Preventing NullReferenceException on WebAuthenticatorIntermediateActivity

### Description

We have a PR [here](https://github.com/dotnet/maui/pull/12347) to address an NRE inside the `WebAuthenticatorIntermediateActivity.Android.cs` but we have some concerns that the fix won't resolve the entire issue.  It doesn't seem like the fix was validated against an actual [repro](https://github.com/dotnet/maui/pull/12347#issuecomment-1412348133). Upon reviewing the code, we feel like this is going to just move the exception down to the `OnResume` call. 

We currently don't have a repro for this so once we're able to get validation that this fix worked on .NET 8 or a repro we can validate then we'll look into back porting. 

### Reproduction

These are from @vhugogarcia 

https://github.com/vhugogarcia/MauiOAuthDemo

https://play.google.com/store/apps/details?id=com.mi.globalbrowser

This is the browser, I used on my Android simulator locally to reproduce the crash. 

Steps:
1. On an Android emulator with Google Play Services installed and Google Play Store configured, search for "Mi Browser" and install it.
2. Open "Mi Browser" app and configure it as default browser in Android.
3. Run the application on the Android device/simulator and try to login with the following attached demo credentials.
4. Expect the crash to happen.


![image](https://user-images.githubusercontent.com/5375137/224114204-aa6cb727-b283-4539-b5ee-b0cfbefa7b25.png)

Follows the error logs that I gets randonly and not very often:

```
[OpenGLRenderer] EglManager::makeCurrent mED = 0x1, surface = 0x0, mEC = 0x76565380, error = EGL_SUCCESS
[0:] System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Microsoft.Maui.Authentication.WebAuthenticatorImplementation.AuthenticateAsync(WebAuthenticatorOptions webAuthenticatorOptions) in D:\a\_work\1\s\src\Essentials\src\WebAuthenticator\WebAuthenticator.android.cs:line 80
   at MyApp.Services.Auth.Browser.<>c__DisplayClass2_0.<<InvokeAsync>b__0>d.MoveNext() in C:\MyApp\src\StudyCompanion\Services\Auth\Browser.cs:line 22
--- End of stack trace from previous location ---
   at MyApp.Services.Auth.Browser.InvokeAsync(BrowserOptions options, CancellationToken cancellationToken) in C:\MyApp\src\StudyCompanion\Services\Auth\Browser.cs:line 20
[0:] 03-22 19:57:27.822 [MyApp.Views.Welcome.WelcomePageViewModel-ERR-1] T: System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Microsoft.Maui.Authentication.WebAuthenticatorImplementation.AuthenticateAsync(WebAuthenticatorOptions webAuthenticatorOptions) in D:\a\_work\1\s\src\Essentials\src\WebAuthenticator\WebAuthenticator.android.cs:line 80
   at MyApp.Services.Auth.Browser.<>c__DisplayClass2_0.<<InvokeAsync>b__0>d.MoveNext() in C:\MyApp\src\StudyCompanion\Services\Auth\Browser.cs:line 22
--- End of stack trace from previous location ---
   at MyApp.Services.Auth.Browser.InvokeAsync(BrowserOptions options, CancellationToken cancellationToken) in C:\MyApp\src\StudyCompanion\Services\Auth\Browser.cs:line 20
 
System.Security.Authentication.AuthenticationException: T: System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Microsoft.Maui.Authentication.WebAuthenticatorImplementation.AuthenticateAsync(WebAuthenticatorOptions webAuthenticatorOptions) in D:\a\_work\1\s\src\Essentials\src\WebAuthenticator\WebAuthenticator.android.cs:line 80
   at MyApp.Services.Auth.Browser.<>c__DisplayClass2_0.<<InvokeAsync>b__0>d.MoveNext() in C:\MyApp\src\StudyCompanion\Services\Auth\Browser.cs:line 22
--- End of stack trace from previous location ---
   at MyApp.Services.Auth.Browser.InvokeAsync(BrowserOptions options, CancellationToken cancellationToken) in C:\MyApp\src\StudyCompanion\Services\Auth\Browser.cs:line 20

   at MyApp.Services.Auth.OidcClientExtensions.Validate[LoginResult](LoginResult result) in C:\MyApp\src\StudyCompanion\Services\Auth\OidcClientExtensions.cs:line 55
   at MyApp.Services.Auth.OidcClientExtensions.ToCredentials(LoginResult loginResult) in C:\MyApp\src\StudyCompanion\Services\Auth\OidcClientExtensions.cs:line 11
   at MyApp.Services.Auth.AuthService.LoginAsync() in C:\MyApp\src\StudyCompanion\Services\Auth\AuthService.cs:line 46
   at MyApp.Views.Welcome.WelcomePageViewModel.Login() in C:\MyApp\src\StudyCompanion\Views\Welcome\WelcomePageViewModel.cs:line 82
[0:category] Application Insights Telemetry: {"name":"AppExceptions","time":"2023-03-22T22:57:27.8877652Z","iKey":"39361284-16bc-4ab2-a792-b3c0fe292ecd","tags":{"ai.device.type":"Phone: Physical","ai.device.id":"motorola","ai.device.osVersion":"Android: 12","ai.device.model":"moto e22","ai.cloud.roleInstance":"localhost.localdomain","ai.user.authUserId":"anonymous@surgent.com","ai.internal.sdkVersion":"dotnetc:2.21.0-429"},"data":{"baseType":"ExceptionData","baseData":{"ver":2,"exceptions":[{"id":959661868,"outerId":0,"typeName":"System.Exception","message":"T: System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Microsoft.Maui.Authentication.WebAuthenticatorImplementation.AuthenticateAsync(WebAuthenticatorOptions webAuthenticatorOptions) in D:\\a\\_work\\1\\s\\src\\Essentials\\src\\WebAuthenticator\\WebAuthenticator.android.cs:line 80
   at MyApp.Services.Auth.Browser.<>c__DisplayClass2_0.<<InvokeAsync>b__0>d.MoveNext() in C:\\MyApp\\src\\StudyCompanion\\Services\\Auth\\Browser.cs:line 22
--- End of stack trace from previous location ---
   at MyApp.Services.Auth.Browser.InvokeAsync(BrowserOptions options, CancellationToken cancellationToken) in C:\\MyApp\\src\\StudyCompanion\\Services\\Auth\\Browser.cs:line 20
","hasFullStack":true},{"id":201167535,"outerId":959661868,"typeName":"System.Security.Authentication.AuthenticationException","message":"T: System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Microsoft.Maui.Authentication.WebAuthenticatorImplementation.AuthenticateAsync(WebAuthenticatorOptions webAuthenticatorOptions) in D:\\a\\_work\\1\\s\\src\\Essentials\\src\\WebAuthenticator\\WebAuthenticator.android.cs:line 80
   at MyApp.Services.Auth.Browser.<>c__DisplayClass2_0.<<InvokeAsync>b__0>d.MoveNext() in C:\\MyApp\\src\\StudyCompanion\\Services\\Auth\\Browser.cs:line 22
--- End of stack trace from previous location ---
   at MyApp.Services.Auth.Browser.InvokeAsync(BrowserOptions options, CancellationToken cancellationToken) in C:\\MyApp\\src\\StudyCompanion\\Services\\Auth\\Browser.cs:line 20
","hasFullStack":true,"parsedStack":[{"level":0,"method":"MyApp.Services.Auth.OidcClientExtensions.Validate","assembly":"StudyCompanion, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null","fileName":"C:\\MyApp\\src\\StudyCompanion\\Services\\Auth\\OidcClientExtensions.cs","line":55},{"level":1,"method":"MyApp.Services.Auth.OidcClientExtensions.ToCredentials","assembly":"StudyCompanion, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null","fileName":"C:\\MyApp\\src\\StudyCompanion\\Services\\Auth\\OidcClientExtensions.cs","line":11},{"level":2,"method":"MyApp.Services.Auth.AuthService+<LoginAsync>d__7.MoveNext","assembly":"StudyCompanion, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null","fileName":"C:\\MyApp\\src\\StudyCompanion\\Services\\Auth\\AuthService.cs","line":46},{"level":3,"method":"System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw","assembly":"System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e","line":0},{"level":4,"method":"System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess","assembly":"System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e","line":0},{"level":5,"method":"System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification","assembly":"System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e","line":0},{"level":6,"method":"System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd","assembly":"System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e","line":0},{"level":7,"method":"System.Runtime.CompilerServices.TaskAwaiter`1[[System.ValueTuple`2[[System.String, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Guid, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].GetResult","assembly":"System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e","line":0},{"level":8,"method":"MyApp.Views.Welcome.WelcomePageViewModel+<Login>d__22.MoveNext","assembly":"StudyCompanion, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null","fileName":"C:\\MyApp\\src\\StudyCompanion\\Views\\Welcome\\WelcomePageViewModel.cs","line":82}]}],"properties":{"BuildId":"1"}}}}
[0:] 03-22 19:57:27.985 [Microsoft.Maui.Controls.BindableObject-WRN-1] Cannot convert Microsoft.Maui.Controls.SolidColorBrush to type 'Microsoft.Maui.Graphics.Color' 
[0:] 03-22 19:57:27.995 [Microsoft.Maui.Controls.BindableObject-WRN-1] Cannot convert Microsoft.Maui.Controls.SolidColorBrush to type 'Microsoft.Maui.Graphics.Color' 
[0:] 03-22 19:57:28.089 [Microsoft.Maui.Controls.BindableObject-WRN-1] Cannot convert Microsoft.Maui.Controls.SolidColorBrush to type 'Microsoft.Maui.Graphics.Color' 
[0:] 03-22 19:57:28.103 [Microsoft.Maui.Controls.BindableObject-WRN-1] Cannot convert Microsoft.Maui.Controls.SolidColorBrush to type 'Microsoft.Maui.Graphics.Color' 
[0:] 03-22 19:57:28.117 [MyApp.Views.Dialog.AlertDialogPageViewModel-FTL-1] AlertDialogPageViewModel.OnInitialize() 
[0:category] Application Insights Telemetry: {"name":"AppTraces","time":"2023-03-22T22:57:28.1234670Z","iKey":"39361284-16bc-4ab2-a792-b3c0fe292ecd","tags":{"ai.device.type":"Phone: Physical","ai.device.id":"motorola","ai.device.osVersion":"Android: 12","ai.device.model":"moto e22","ai.cloud.roleInstance":"localhost.localdomain","ai.user.authUserId":"anonymous@surgent.com","ai.internal.sdkVersion":"dotnetc:2.21.0-429"},"data":{"baseType":"MessageData","baseData":{"ver":2,"message":"AlertDialogPageViewModel.OnInitialize()","severityLevel":"Verbose","properties":{"BuildId":"1"}}}}
[0:] 03-22 19:57:28.129 [MyApp.Views.Dialog.AlertDialogPageViewModel-FTL-1] AlertDialogPageViewModel.OnNavigatedTo() 
[0:category] Application Insights Telemetry: {"name":"AppTraces","time":"2023-03-22T22:57:28.1333751Z","iKey":"39361284-16bc-4ab2-a792-b3c0fe292ecd","tags":{"ai.device.type":"Phone: Physical","ai.device.id":"motorola","ai.device.osVersion":"Android: 12","ai.device.model":"moto e22","ai.cloud.roleInstance":"localhost.localdomain","ai.user.authUserId":"anonymous@surgent.com","ai.internal.sdkVersion":"dotnetc:2.21.0-429"},"data":{"baseType":"MessageData","baseData":{"ver":2,"message":"AlertDialogPageViewModel.OnNavigatedTo()","severityLevel":"Verbose","properties":{"BuildId":"1"}}}}
**Java.Lang.IllegalStateException:** 'Can not perform this action after onSaveInstanceState'

**Java.Lang.IllegalStateException:** 'Can not perform this action after onSaveInstanceState'

[mono-rt] [ERROR] FATAL UNHANDLED EXCEPTION: Java.Lang.IllegalStateException: Can not perform this action after onSaveInstanceState
[mono-rt]    at Java.Interop.JniEnvironment.InstanceMethods.CallIntMethod(JniObjectReference instance, JniMethodInfo method, JniArgumentValue* args) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/obj/Release/net7.0/JniEnvironment.g.cs:line 11725
[mono-rt]    at Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeAbstractInt32Method(String encodedMember, IJavaPeerable self, JniArgumentValue* parameters) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniPeerMembers.JniInstanceMethods_Invoke.cs:line 492
[mono-rt]    at AndroidX.Fragment.App.BackStackRecord.Commit() in /Users/runner/work/1/s/generated/androidx.fragment.fragment/obj/Release/net6.0-android/generated/src/AndroidX.Fragment.App.BackStackRecord.cs:line 131
[mono-rt]    at Microsoft.Maui.Controls.Platform.ModalNavigationManager.ModalContainer..ctor(IMauiContext windowMauiContext, Page modal, ViewGroup parentView) in D:\a\_work\1\s\src\Controls\src\Core\Platform\ModalNavigationManager\ModalNavigationManager.Android.cs:line 228
[mono-rt]    at Microsoft.Maui.Controls.Platform.ModalNavigationManager.PresentModal(Page modal, Boolean animated) in D:\a\_work\1\s\src\Controls\src\Core\Platform\ModalNavigationManager\ModalNavigationManager.Android.cs:line 130
[mono-rt]    at Microsoft.Maui.Controls.Platform.ModalNavigationManager.PushModalAsync(Page modal, Boolean animated) in D:\a\_work\1\s\src\Controls\src\Core\Platform\ModalNavigationManager\ModalNavigationManager.Android.cs:line 119
[mono-rt]    at Microsoft.Maui.Controls.Window.NavigationImpl.OnPushModal(Page modal, Boolean animated) in D:\a\_work\1\s\src\Controls\src\Core\HandlerImpl\Window\Window.Impl.cs:line 709
[mono-rt]    at Microsoft.Maui.Controls.Shell.NavigationImpl.OnPushModal(Page modal, Boolean animated) in D:\a\_work\1\s\src\Controls\src\Core\Shell\Shell.cs:line 1689
[mono-rt]    at Microsoft.Maui.Controls.ShellSection.GoToAsync(ShellNavigationRequest request, ShellRouteParameters queryData, IServiceProvider services, Nullable`1 animate, Boolean isRelativePopping) in D:\a\_work\1\s\src\Controls\src\Core\Shell\ShellSection.cs:line 566
[mono-rt]    at Microsoft.Maui.Controls.ShellNavigationManager.GoToAsync(ShellNavigationParameters shellNavigationParameters, ShellNavigationRequest navigationRequest) in D:\a\_work\1\s\src\Controls\src\Core\Shell\ShellNavigationManager.cs:line 196
[mono-rt]    at MyApp.Services.Navigation.NavigationService.GoTo(String route, IDictionary`2 args) in C:\MyApp\src\StudyCompanion\Services\Navigation\NavigationService.cs:line 14
[mono-rt]    at MyApp.Services.Dialog.DialogService.<>c__DisplayClass2_0.<<Alert>b__0>d.MoveNext() in C:\MyApp\src\StudyCompanion\Services\Dialog\DialogService.cs:line 30
[mono-rt] --- End of stack trace from previous location ---
[mono-rt]    at MyApp.Services.Dialog.DialogService.Alert(String message, String title, String buttonText) in C:\MyApp\src\StudyCompanion\Services\Dialog\DialogService.cs:line 28
[mono-rt]    at MyApp.Views.Welcome.WelcomePageViewModel.Login() in C:\MyApp\src\StudyCompanion\Views\Welcome\WelcomePageViewModel.cs:line 119
[mono-rt]    at Reactive.Bindings.AsyncReactiveCommand.<>c__DisplayClass6_0.<<Subscribe>b__0>d.MoveNext()
[mono-rt] --- End of stack trace from previous location ---
[mono-rt]    at Reactive.Bindings.AsyncReactiveCommand`1.<ExecuteAsync>d__17[[System.Object, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
[mono-rt]    at Reactive.Bindings.AsyncReactiveCommand`1.<Execute>d__16[[System.Object, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
[mono-rt]    at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
[mono-rt]    at Android.App.SyncContext.<>c__DisplayClass2_0.<Post>b__0() in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.App/SyncContext.cs:line 36
[mono-rt]    at Java.Lang.Thread.RunnableImplementor.Run() in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Thread.cs:line 36
[mono-rt]    at Java.Lang.IRunnableInvoker.n_Run(IntPtr jnienv, IntPtr native__this) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net7.0/android-33/mcw/Java.Lang.IRunnable.cs:line 84
[mono-rt]    at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PP_V(_JniMarshal_PP_V callback, IntPtr jnienv, IntPtr klazz) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 22
[mono-rt]   --- End of managed Java.Lang.IllegalStateException stack trace ---
[mono-rt] java.lang.IllegalStateException: Can not perform this action after onSaveInstanceState
[mono-rt] 	at androidx.fragment.app.FragmentManager.checkStateLoss(FragmentManager.java:1610)
[mono-rt] 	at androidx.fragment.app.FragmentManager.enqueueAction(FragmentManager.java:1650)
[mono-rt] 	at androidx.fragment.app.BackStackRecord.commitInternal(BackStackRecord.java:341)
[mono-rt] 	at androidx.fragment.app.BackStackRecord.commit(BackStackRecord.java:306)
[mono-rt] 	at crc6468b6408a11370c2f.WebAuthenticatorIntermediateActivity.n_onResume(Native Method)
[mono-rt] 	at crc6468b6408a11370c2f.WebAuthenticatorIntermediateActivity.onResume(WebAuthenticatorIntermediateActivity.java:41)
[mono-rt] 	at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1476)
[mono-rt] 	at android.app.Activity.performResume(Activity.java:8190)
[mono-rt] 	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4787)
[mono-rt] 	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4831)
[mono-rt] 	at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:54)
[mono-rt] 	at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
[mono-rt] 	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
[mono-rt] 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
[mono-rt] 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2247)
[mono-rt] 	at android.os.Handler.dispatchMessage(Handler.java:106)
[mono-rt] 	at android.os.Looper.loopOnce(Looper.java:201)
[mono-rt] 	at android.os.Looper.loop(Looper.java:288)
[mono-rt] 	at android.app.ActivityThread.main(ActivityThread.java:7881)
[mono-rt] 	at java.lang.reflect.Method.invoke(Native Method)
[mono-rt] 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:568)
[mono-rt] 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1045)
[mono-rt] 
[mono-rt]   --- End of managed Java.Lang.IllegalStateException stack trace ---
[mono-rt] java.lang.IllegalStateException: Can not perform this action after onSaveInstanceState
[mono-rt] 	at androidx.fragment.app.FragmentManager.checkStateLoss(FragmentManager.java:1610)
[mono-rt] 	at androidx.fragment.app.FragmentManager.enqueueAction(FragmentManager.java:1650)
[mono-rt] 	at androidx.fragment.app.BackStackRecord.commitInternal(BackStackRecord.java:341)
[mono-rt] 	at androidx.fragment.app.BackStackRecord.commit(BackStackRecord.java:306)
[mono-rt] 	at crc6468b6408a11370c2f.WebAuthenticatorIntermediateActivity.n_onResume(Native Method)
[mono-rt] 	at crc6468b6408a11370c2f.WebAuthenticatorIntermediateActivity.onResume(WebAuthenticatorIntermediateActivity.java:41)
[mono-rt] 	at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1476)
[mono-rt] 	at android.app.Activity.performResume(Activity.java:8190)
[mono-rt] 	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4787)
[mono-rt] 	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4831)
[mono-rt] 	at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:54)
[mono-rt] 	at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
[mono-rt] 	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
[mono-rt] 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
[mono-rt] 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2247)
[mono-rt] 	at android.os.Handler.dispatchMessage(Handler.java:106)
[mono-rt] 	at android.os.Looper.loopOnce(Looper.java:201)
[mono-rt] 	at android.os.Looper.loop(Looper.java:288)
[mono-rt] 	at android.app.ActivityThread.main(ActivityThread.java:7881)
[mono-rt] 	at java.lang.reflect.Method.invoke(Native Method)
[mono-rt] 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:568)
[mono-rt] 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1045)
[mono-rt] 
[libc] FORTIFY: pthread_mutex_lock called on a destroyed mutex (0x9e83ca60)

```

--- a/src/Essentials/src/WebAuthenticator/WebAuthenticator.android.cs
+++ b/src/Essentials/src/WebAuthenticator/WebAuthenticator.android.cs
@@ -13,6 +13,8 @@ namespace Microsoft.Maui.Authentication
 		Uri currentRedirectUri = null;
 		WebAuthenticatorOptions currentOptions = null;
 
+		internal bool AuthenticatingWithCustomTabs { get; private set; } = false;
+
 		public bool OnResumeCallback(Intent intent)
 		{
 			// If we aren't waiting on a task, don't handle the url
@@ -69,8 +71,12 @@ namespace Microsoft.Maui.Authentication
 
 			tcsResponse = new TaskCompletionSource<WebAuthenticatorResult>();
 			currentRedirectUri = callbackUrl;
+			
+			// Try to start with custom tabs if the system supports it and we resolve it
+			AuthenticatingWithCustomTabs = await StartCustomTabsActivity(url);
 
-			if (!(await StartCustomTabsActivity(url)))
+			// Fall back to using the system browser if necessary
+			if (!AuthenticatingWithCustomTabs)
 			{
 				// Fall back to opening the system-registered browser if necessary
 				var urlOriginalString = url.OriginalString;

--- a/src/Essentials/src/WebAuthenticator/WebAuthenticator.android.cs
+++ b/src/Essentials/src/WebAuthenticator/WebAuthenticator.android.cs
@@ -71,7 +71,7 @@ namespace Microsoft.Maui.Authentication
 
 			tcsResponse = new TaskCompletionSource<WebAuthenticatorResult>();
 			currentRedirectUri = callbackUrl;
-			
+
 			// Try to start with custom tabs if the system supports it and we resolve it
 			AuthenticatingWithCustomTabs = await StartCustomTabsActivity(url);
 

--- a/src/Essentials/src/WebAuthenticator/WebAuthenticator.shared.cs
+++ b/src/Essentials/src/WebAuthenticator/WebAuthenticator.shared.cs
@@ -102,6 +102,11 @@ namespace Microsoft.Maui.Authentication
 			return platform;
 		}
 
+#if ANDROID
+		internal static bool IsAuthenticatingWithCustomTabs(this IWebAuthenticator webAuthenticator)
+			=> (webAuthenticator as WebAuthenticatorImplementation)?.AuthenticatingWithCustomTabs ?? false;
+#endif
+
 		/// <summary>
 		/// Begin an authentication flow by navigating to the specified url and waiting for a callback/redirect to the callbackUrl scheme.
 		/// </summary>

--- a/src/Essentials/src/WebAuthenticator/WebAuthenticatorCallbackActivity.android.cs
+++ b/src/Essentials/src/WebAuthenticator/WebAuthenticatorCallbackActivity.android.cs
@@ -24,7 +24,7 @@ namespace Microsoft.Maui.Authentication
 				// No intermediate activity if we returned from a system browser
 				// intent since there's no custom tab instance to clean up
 				WebAuthenticator.Default.OnResume(Intent);
-			} 
+			}
 
 			// finish this activity
 			Finish();

--- a/src/Essentials/src/WebAuthenticator/WebAuthenticatorCallbackActivity.android.cs
+++ b/src/Essentials/src/WebAuthenticator/WebAuthenticatorCallbackActivity.android.cs
@@ -10,11 +10,21 @@ namespace Microsoft.Maui.Authentication
 		{
 			base.OnCreate(savedInstanceState);
 
-			// start the intermediate activity again with flags to close the custom tabs
-			var intent = new Intent(this, typeof(WebAuthenticatorIntermediateActivity));
-			intent.SetData(Intent.Data);
-			intent.AddFlags(ActivityFlags.ClearTop | ActivityFlags.SingleTop);
-			StartActivity(intent);
+			// Check how we launched the flow initially
+			if (WebAuthenticator.Default.IsAuthenticatingWithCustomTabs())
+			{
+				// start the intermediate activity again with flags to close the custom tabs
+				var intent = new Intent(this, typeof(WebAuthenticatorIntermediateActivity));
+				intent.SetData(Intent.Data);
+				intent.AddFlags(ActivityFlags.ClearTop | ActivityFlags.SingleTop);
+				StartActivity(intent);
+			}
+			else
+			{
+				// No intermediate activity if we returned from a system browser
+				// intent since there's no custom tab instance to clean up
+				WebAuthenticator.Default.OnResume(Intent);
+			} 
 
 			// finish this activity
 			Finish();

